### DISCUSSION TITLE
Renovate is using `CHANGELOG.json` when `CHANGELOG.md` also exists

### DISCUSSION THREAD

**Discussion by frantic1048** [renovatebot/renovate#25828] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.59.8

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

## Symptom

We have some npm packages managed with beachball[^1], for each release, it generates both formats of changelog under package path:

- `CHANGELOG.json`: JSON file, changelog data for programmatical usage
- `CHANGELOG.md`: Markdown file, human readable changelog

For example: https://github.com/RightCapitalHQ/frontend-style-guide/tree/3fe7eea47b45ecd796340dc5ef1f98d034fd1c25/packages/eslint-config-javascript

When Renovate is generating Pull Request for such package, Renovate will report:

```
DEBUG: Multiple candidates for changelog file, using packages/eslint-config-javascript/CHANGELOG.json (repository=frantic1048/renovate-uses-json-changelog-while-md-exists, branch=action-renovate/rightcapital-eslint-config-javascript-14.x)
```

And generates empty release notes for the package. https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/pull/2

Example:

- Repo: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists
- Renovate config: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/blob/main/renovate.json
- Renovate self-hosted config: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/blob/main/renovate-config.js
- Renovate usage: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/blob/main/.github/workflows/main.yml#L8
- Renovate Log: https://github.com/frantic1048/renovate-uses-json-changelog-while-md-exists/actions/runs/6888285336/job/18736975879#step:3:537

## Cause

In [`github.getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/github/index.ts#L22) (called from [`getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/release-notes.ts#L237)), Renovate is [selecting the first file](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/github/index.ts#L63) filtered by [`changelogFilenameRegex`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/github/index.ts#L51), when multiple capable changelog files are found.

Unfortunately, `CHANGELOG.md` and `CHANGELOG.json` both pass the test of `changelogFilenameRegex`, resulting above symptom.

## Possible fix

Given the fact that Renovate is trying to [read changelog as a Markdown file](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/release-notes.ts#L216) to extract useful info.

I think Renovate may give markdown file a higher priority during selecting a changelog from multiple choices.

## Other info

[`gitlab.getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/update/pr/changelog/gitlab/index.ts#L17) and [`bitbucket.getReleaseNotesMd()`](https://github.com/renovatebot/renovate/blob/b3424c626cb8e44b922b91f669826a466ccee5c8/lib/workers/repository/update/pr/changelog/bitbucket/index.ts#L17) have similar logic and could suffer from same symptom 🤔 

[^1]: a tool to automating version bump and package publish for node projects, see: https://microsoft.github.io/beachball/

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Multiple candidates for changelog file, using packages/eslint-config-javascript/CHANGELOG.json (repository=frantic1048/renovate-uses-json-changelog-while-md-exists, branch=action-renovate/rightcapital-eslint-config-javascript-14.x)
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment, marked as answer]
Converted to bug issue: #25830
